# Ejercicios 4

#### 1 Ejercicio

* Crear un array de tamaño 20 donde todos los elementos estén inicializados a cero. Por favor llámalo __a__.

In [2]:
# Sol:
import numpy as np
a=np.zeros(20)
a


array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.])

* Escribe una expresión Python para cambiar el valor de  los primeros 5 elementos; el nuevo valor es 10.

In [4]:
# Sol:
a[0:5]=10
a

array([ 10.,  10.,  10.,  10.,  10.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.])

* Modifica los siguientes 10 valores. Los nuevos valores serán el resultado de la secuencia de números pares comenzando por el 12. Utiliza la función __arange__.

In [6]:
# Sol:
a[5:15]=np.arange(12,32,2)
a

array([ 10.,  10.,  10.,  10.,  10.,  12.,  14.,  16.,  18.,  20.,  22.,
        24.,  26.,  28.,  30.,   0.,   0.,   0.,   0.,   0.])

* Modifica el valor de los últimos 5 elementos. Su nuevo valor es 30.

In [8]:
# Sol:
a[-5:]=30
a


array([ 10.,  10.,  10.,  10.,  10.,  12.,  14.,  16.,  18.,  20.,  22.,
        24.,  26.,  28.,  30.,  30.,  30.,  30.,  30.,  30.])

#### 2 Ejercicio

El archivo de texto [holland_temperature.dat](./datos/holland_temperature.dat) recoge datos de las diferentes temperaturas registradas en Holanda en los últimos 12 meses.

---
a) Calcula la media de las temperaturas.
   * Descarga el fichero [holland_temperature.dat](./datos/holland_temperature.dat) en tu directorio de trabajo.
   * Carga los datos en una array llamado __temperatura__ mediante la función __loadtxt__.
   * Utiliza la función __mean__ para calcular la media.
    
__Nota__: La media es 10.125.

In [13]:
# Sol
temperatura= np.loadtxt("./datos/holland_temperature.dat")
temperatura

array([  3.1,   3.3,   6.2,   9.2,  13.1,  15.6,  17.9,  17.5,  14.5,
        10.7,   6.7,   3.7])

In [15]:
np.mean(temperatura)

10.125

------
b) Nos interesa conocer los meses del año donde la temperatura ha sido superior a la media.
   * Crear un array de meses [1,...,12]. Usa la función __arange__. 
   * Escribir la expresión que devuelve los meses donde la temperatura ha sido superior a la media usando como máscara un array de booleanos.
    
__Nota__: La solución es: [5,  6,  7,  8,  9, 10]

In [16]:
# Sol:
meses=np.arange(1,13)
meses [temperatura>temperatura.mean()]

array([ 5,  6,  7,  8,  9, 10])

-------
c) Queremos saber qué mes ha estado más cerca de la media. 
* Para ello calcula el array de diferencias con respecto a la media (operación diferencia). 
* Luego utiliza la función __argmin__ de Numpy para calcular el índice del array que contiene el mínimo valor.

__Nota__: la solución es: 'El mes más cercano a los 10.12 grados es el mes 9'

In [17]:
# Sol:
x=temperatura-temperatura.mean()
x


array([-7.025, -6.825, -3.925, -0.925,  2.975,  5.475,  7.775,  7.375,
        4.375,  0.575, -3.425, -6.425])

In [18]:
y=abs(x)
y

array([ 7.025,  6.825,  3.925,  0.925,  2.975,  5.475,  7.775,  7.375,
        4.375,  0.575,  3.425,  6.425])

In [19]:
min(y)
np.argmin(y) 

9

---------------------